### PR TITLE
PBI-111 Backend Integration with RTK

### PR DESCRIPTION
**Pull request by FreeJ1nG**
## Describe your changes
- Added unit tests for vendor page (with proper data mocking with [MSW](https://mswjs.io/docs/getting-started))
- Setup RTK Query for `/vendor` alongside other params that it consumes (with proper caching)
- Add global loading overlay component that can be accessed via `useCommonStore`
- Clean up things relating to vendor page that were still using dummy data

## Issue ticket number and link
- [ADU-133](https://linear.app/adudu/issue/ADU-133/integrate-msw-for-unit-testing-rtk-query)
- [ADU-103](https://linear.app/adudu/issue/ADU-103/view-all-vendor-integration)
- [ADU-104](https://linear.app/adudu/issue/ADU-104/filter-vendor-integration)

### CODE REVIEW

File: src/components/pages/vendor.tsx
Comment (FreeJ1nG) [inline src/components/pages/vendor.tsx:1]: For anyone else trying to use RTK Query, this is the example file to look at

File: src/components/pages/vendor.tsx [hunk 1]
```diff
@@ -1,23 +1,33 @@
...
+        ...result,
+        vendors: result.data?.Vendors,
+        metadata: result.data?.Metadata,
+      }),
+    },
+  )
+
+  useEffect(() => {
+    setShowLoadingOverlay(!isSuccess)
+  }, [isSuccess, setShowLoadingOverlay])
```
Comment (FreeJ1nG) [inline src/components/pages/vendor.tsx:29]: Utility to show global loading overlay can be seen here (with `data-testid` = `loading-overlay`, this can be used when you want to check if the data is no longer loading)

File: src/lib/msw/vendor.ts
Comment (FreeJ1nG) [inline src/lib/msw/vendor.ts:1]: This file shows MSW, which intercepts any requests that our tests tried to make, and returns a response (a mock response in this case, that aligns with the backend's data definitions)

Read the docs for more info: https://mswjs.io/docs/getting-started

File: src/lib/redux/features/vendor/api.ts
Comment (FreeJ1nG) [inline src/lib/redux/features/vendor/api.ts:1]: This file is how we build queries/mutations with RTK Query, `providesTags` and `invalidatesTags` is important for caching

Read the docs for more info of tags: https://redux-toolkit.js.org/rtk-query/usage/automated-refetching

(For what it's worth, let's not care so much about caching for now, you can inquire me for more details on this)

File: src/lib/redux/services/api.ts [hunk 2]
```diff
@@ -1,13 +1,17 @@
...
+  baseQuery: baseQueryWithZodValidation(
+    fetchBaseQuery({
+      baseUrl: import.meta.env.VITE_API_BASE_URL,
+      async prepareHeaders() {
+        // TODO: Do JWT Fetching logic and attach it to the header here
+      },
+    }),
+  ),
+  tagTypes: ['Vendor'],
   endpoints: () => ({}),
```
Comment (FreeJ1nG) [inline src/lib/redux/services/api.ts]: I decided to use zod for validating the backend's data shape

File: src/lib/redux/services/baseQuery.ts
Comment (FreeJ1nG) [inline src/lib/redux/services/baseQuery.ts:1]: Custom base query with zod to validate the data response from the backend

File: src/lib/testing/utils.tsx
Comment (FreeJ1nG) [inline src/lib/testing/utils.tsx:1]: I made some changes to the `withWrappers` utility, you can read the JSDoc I wrote to understand it better

TL;DR: If `withRoot` is set to `true`, then The root of the app will be rendered (Navbar, LoadingOverlay utility, etc. This can be seen in the `routes/__root.tsx` file)

File: src/lib/zustand/common.ts
Comment (FreeJ1nG) [inline src/lib/zustand/common.ts:1]: This is the store we'll use to do the loading overlay logic (this can be accessed globally as long as it's within react's lifecycle)

File: src/components/pages/__tests__/vendor.test.tsx [hunk 3]
```diff
@@ -16,9 +16,14 @@ describe('<VendorPage/>', () => {
...
-    render(withWrappers(<VendorPage />))
-    expect(screen.getByRole('table')).toBeInTheDocument()
+  it('should render the vendors table content properly', async () => {
+    const { container } = render(
+      withWrappers(<VendorPage />, { withRoot: true }),
+    )
+    await waitFor(async () => {
+      expect(screen.queryByTestId('loading-overlay')).not.toBeInTheDocument()
+    })
+    expect(container.innerText).toMatchSnapshot()
```
Comment (bryannaufal) [inline src/components/pages/__tests__/vendor.test.tsx:26]: Why did you choose to use `toMatchSnapshot` instead of `toMatchInlineSnapshot` here?

Comment (FreeJ1nG) [reply src/components/pages/__tests__/vendor.test.tsx:26]: `toMatchInlineSnapshot` will store the snapshot data inside of the function itself

For example, if the snapshot evaluates to `"Hi there, how are you doing?"`, then it'd be something like
```
expect(...).toMatchInlineSnapshot(`"Hi there, how are you doing?"`)
```
This behavior might not be ideal when the snapshot might be large, multiline text, which it might be in this case, so I decided to use `.toMatchSnapshot`, which stores the snapshot in a `__snapshots__` folder instead of storing it inline
